src/backend/auxil/range_alloc.rs

```diff
@@ -0,0 +1,169 @@
+use std::ops::{Add, AddAssign, Range, Sub};
+use std::fmt::Debug;
+
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub struct RangeAllocator<T> {
+    /// The range this allocator covers.
+    range: Range<T>,
+    /// A Vec of ranges in this heap which are unused.
+    /// Must be ordered with ascending range start to permit short circuiting allocation.
+    /// No two ranges in this vec may overlap.
+    empty_ranges: Vec<Range<T>>,
+}
+
+impl<T> RangeAllocator<T>
+    where T: Clone + Copy + Add<Output=T> + AddAssign + Sub<Output=T> + Eq + PartialOrd + Debug
+{
+    pub fn new(range: Range<T>) -> Self {
+        RangeAllocator {
+            range: range.clone(),
+            empty_ranges: vec![range.clone()],
+        }
+    }
+
+    pub fn range(&self) -> Range<T> {
+        self.range.clone()
+    }
+
+    pub fn allocate_range(&mut self, length: T) -> Option<Range<T>> {
+        let mut best_range_tuple: Option<(usize, Range<T>)> = None;
+        for range in self.empty_ranges.iter().cloned().enumerate() {
+            if range.1.end - range.1.start < length {
```

Comment:
nit: start + length < end might be easier to read

Reply:
Or calculate the size of the current range once and store it in a variable. Seem to be recalculated multiple times below